whois-api/src/test/java/net/ripe/db/whois/api/rest/WhoisRestServiceDoSTestIntegration.java

```diff
@@ -0,0 +1,249 @@
+            "auth:        SSO person@net.net\n" +
+            "mnt-by:      OWNER-MNT\n" +
+            "source:      TEST");
+
+    private static final Integer PENULTIMATE_OK_UPDATE_POSITION = 8;
+
+    private static final Integer PENULTIMATE_OK_LOOKUP_POSITION = 48;
+
+    private static final Integer MAXIMUM_UPDATE_REQUESTS_ALLOWED_PER_SECOND = 10;
+
+    private static final Integer MAXIMUM_LOOKUP_REQUESTS_ALLOWED_PER_SECOND = 50;
+
+    private static final Integer SECONDS_NEEDED_TO_FREE_IP = 1;
+
+    @BeforeAll
+    public static void beforeClass() {
+        System.setProperty("dos.filter.enabled", "true");
+    }
+
+    @BeforeEach
+    public void setup() {
+        databaseHelper.addObject("person: Test Person\nnic-hdl: TP1-TEST");
+        databaseHelper.addObject(OWNER_MNT);
+        testDateTimeProvider.setTime(LocalDateTime.parse("2001-02-04T17:00:00"));
+    }
+
+    // lookup
+    @Test
+    public void multiple_lookup_per_second_then_429_too_many_requests() throws InterruptedException {
+        final Response response = RestTest.target(getPort(), "whois/test/mntner/OWNER-MNT?clientIp=10.20.30.40")
+                .request()
+                .get();
+        assertThat(HttpStatus.OK_200, is(response.getStatus()));
+
+        final Invocation.Builder lookupRequest = RestTest.target(getPort(), "whois/test/mntner/OWNER-MNT?clientIp=10.20.30.40").request();
+
+        // Simulate a DoS attack by sending many GET requests in a short time
+        final List<Integer> responsesCodes = IntStream.range(0, MAXIMUM_LOOKUP_REQUESTS_ALLOWED_PER_SECOND)
+                .mapToObj(lookupCount -> lookupRequest.get())
+                .map(Response::getStatus)
+                .toList();
+
+        assertThat(HttpStatus.OK_200, is(responsesCodes.get(PENULTIMATE_OK_LOOKUP_POSITION)));
+        assertThat(HttpStatus.TOO_MANY_REQUESTS_429, is(responsesCodes.getLast()));
+
+        TimeUnit.SECONDS.sleep(SECONDS_NEEDED_TO_FREE_IP); // Free the IP after one second
+
+        //After a second, the user can perform more requests
+        assertThat(HttpStatus.OK_200, is(lookupRequest.get().getStatus()));
+    }
+
+    @Test
+    public void multiple_async_lookup_per_second_then_429_too_many_requests() throws InterruptedException {
+        final Response creation = RestTest.target(getPort(), "whois/test/mntner/OWNER-MNT?clientIp=10.20.30.40")
+                .request()
+                .get();
+        assertThat(HttpStatus.OK_200, is(creation.getStatus()));
+
+        final Invocation.Builder lookupRequest = RestTest.target(getPort(), "whois/test/mntner/OWNER-MNT?clientIp=10.20.30.40").request();
+
+        // Simulate a DoS attack by sending many GET requests in a short time asynchronously
+        final List<Integer> responsesCodes = IntStream.range(0, MAXIMUM_LOOKUP_REQUESTS_ALLOWED_PER_SECOND)
+                .parallel()
+                .mapToObj(lookupCount -> lookupRequest.get())
+                .map(Response::getStatus)
```

Comment:
use groupBy so you can get success and failed  count in one query

Reply:
yes, good idea

Reply:
Done!